Adds a sans-io fsm to the read operation.

mzimbres opened the pull request:
@anarthal: Notes for the code review
   - Adds a fsm for read operation following your work on `exec_fsm`.
   - Log every event of the `reader_fsm` after it resumes, which improves log information content.
   - The logger interface has to be extended with types that are actually in the `detail` namespace, so I make fsm actions a public type, this is needed to interpret log messages. I would like to do this to the `exec_fsm` too once a log object becomes available for `async_exec`, so I did not change your code yet.
   - Cancellation hasn't been addresses yet for read operation. This will need more thought and a separate PR.

Example log for the subscriber example. Once all ops are refactored to use an fsm, the log will become very detailed and more helpful
```
(Boost.Redis) resolve results: 127.0.0.1:6379
(Boost.Redis) connected to 127.0.0.1:6379
(Boost.Redis) fsm action: (reader_action::type::setup_cancellation, 0, Success)
(Boost.Redis) fsm action: (reader_action::type::append_some, 0, Success)
(Boost.Redis) writer_op: 117 bytes written.
(Boost.Redis) fsm action: (reader_action::type::notify_push_receiver, 36, Success)
(Boost.Redis) fsm action: (reader_action::type::append_some, 0, Success)
(Boost.Redis) hello_op: Success
subscribe channel 1
^C
(Boost.Redis) ping_op (4): Operation canceled
(Boost.Redis) check_timeout_op (2): Operation canceled
(Boost.Redis) fsm action: (reader_action::type::cancel_run, 0, Success)
(Boost.Redis) fsm action: (reader_action::type::done, 0, Operation canceled)
(Boost.Redis) writer_op (3): connection is closed.
```

anarthal commented on include/boost/redis/connection.hpp:
```diff
@@ -206,85 +207,56 @@ struct writer_op {
-               self.complete(ec);
+         switch (act.type_) {
+            case reader_action::type::setup_cancellation:
+               self.reset_cancellation_state(asio::enable_total_cancellation());
```
Maybe partial cancellation makes more sense here? In any case, we'll handle this later.

anarthal commented on test/test_reader_fsm.cpp:
```diff
@@ -0,0 +1,150 @@
+   test_push();
+   test_read_error();
+   test_parse_error();
+
```
I'd add the following cases:

* An incomplete message is delivered (i.e. it's fragmented in two reads).
* The channel notification fails

anarthal commented on include/boost/redis/impl/logger.ipp:
```diff
@@ -89,19 +89,15 @@ void logger::on_write(system::error_code const& ec, std::string_view payload)
-      std::clog << "reader_op: " << n << " bytes read.";
-
-   std::clog << std::endl;
+   std::clog << "fsm action: (" << to_string(action.type_) << ", " << action.push_size_ << ", "
```
I've got the concern that we might be logging too much for the `level::info` level. It looks to me that most of this info really belongs to `level::debug`. However, fine-tuning this is probably better done once we merge the logger interface PR, because we can start moving logging to fsms (and hence testing it).

mzimbres commented:
I think I addressed all your points. Please have a look at it again. Are you about to get the log PR merged? I will wait and then rebase on top of your changes so that I can remove the usage of types in `detail` in the logger interface.
